Add support for vibe.d

jpf91 opened the pull request:
This pull request adds support for vibe.d as a backend.

I had to do some refactoring to make this possible. I had to move errno checking into the Stream classes and it became obvious that dlang-requests is quite inconsistent when error checking. Some calls do enforce timeouts, some don't. Some calls throw exceptions, some don't. I tried to keep the error handling behaviour unchanged, but I recommend to unify error handling later on (e.g. always propagate TimeoutExceptions: https://github.com/ikod/dlang-requests/commit/ee7e64e506a9f14601f593bb81db76aea9cdadcf#diff-5afa5611d415deda31281152a84a7b8aL109).

Note:
* vibe.d does not support write timeouts or timeouts when connecting. See https://github.com/rejectedsoftware/vibe.d/issues/1137
* I'm not sure if the timeout implementation for SSL is 100% bullet proof. By calling `waitForData` first we enforce the timeout. But if the data received is only TLS management data `leastSize` might block, as the stream is not empty (so it can't return 0) and it can't return a value > 0 as there's no decrypted data. So it'll probably read data from the stream, without enforcing the timeout. OTOH there's no better way to implement the timeout (there's the `readTimeout` property, but this will close the connection and make it difficult to throw an TimeoutException. Additionally the timeout semantics are different as well: https://github.com/rejectedsoftware/vibe.d/issues/622).

jpf91 commented:
I've opened a vibe.d forum thread asking about read timeout details:
http://forum.rejectedsoftware.com/groups/rejectedsoftware.vibed/thread/32752/

ikod commented:
Hello!
Excelent! I'll check it ASAP.
Few questions:
Can you, please, show simple code example on using requests with vibe.d so that I can put it to README.md?
I didn't find unittests for vibe.d version in diffs, and we need them.

jpf91 commented:
> Hello!

Hi :-) Thanks for writing this nice library BTW :+1: 

> This special handling of EINTR on Posix platform should be preserved.

Yes, that makes sense. I've just moved that code into `SocketStream.receive` so it still works as before: https://github.com/jpf91/dlang-requests/blob/9bba119c01040bed205488c9b574bec40b1b7eee/source/requests/streams.d#L833

> Can you, please, show simple code example on using requests with vibe.d so that I can put it to README.md?

The way I've implemented this the `vibe.d/socket` backends are chosen at compile time. See https://github.com/jpf91/dlang-requests/blob/vibe/source/requests/streams.d#L988 the `TCPStream` and `SSLStream` aliases are either set to `TCPVibeStream`, `SSLVibeStream` or `TCPSocketStream` `SSLSocketStream`. The http/ftp code then only uses the `TCPStream` and `SSLStream` aliases.

The user simply uses the correct configuration:

```json
"dependencies": {
    "requests": "~>0.1.9"
},
"subConfigurations": {
    "requests": "vibed"
}
```
and now all calls to the standard API use vibe.d as a backend. I've already added that to the README in this pull request.

This way there's no need to change any user code. The only problem is that it's not possible to have support for std.socket and vibe.d in the same application. But I don't think we really need that feature. 

> I didn't find unittests for vibe.d version in diffs, and we need them.

By running `dub test --config=vibed` all existing unittests are run with the vibe.d backend. I've also updated the `.travis.yml` file in the pull request to run these tests:
https://github.com/jpf91/dlang-requests/commit/9bba119c01040bed205488c9b574bec40b1b7eee
This is another benefit of simply selecting the 'backend' at compile time.

ikod commented:
> The way I've implemented this the vibe.d/socket backends are chosen at compile time. See https://github.com/jpf91/dlang-requests/blob/vibe/source/requests/streams.d#L988 the TCPStream and SSLStream aliases are either set to TCPVibeStream, SSLVibeStream or TCPSocketStream SSLSocketStream. The http/ftp code then only uses the TCPStream and SSLStream aliases.

I'm not very familiar with vibe.d, so forgive me my stupid questions.
Let we consider case, when user would like to use Requests in the vibe.d app. Using using plain old sockets in vibe.d prohibited because it will block current thread/fiber. Using vibe.d sockets under Requests allow to safely use "synchonous" Requests API in vibe.d app. Is it correct?

jpf91 commented:
> Using vibe.d sockets under Requests allow to safely use "synchonous" Requests API in vibe.d app. Is it correct?

Yes, that's correct. The std.socket API always blocks a complete thread, the vibe.d API only blocks the current Fiber. Here's a more advanced example:

```d
import requests, vibe.d;

shared static this()
{
    void taskMain()
    {
        logInfo("Task created");
        auto r1 = getContent("http://httpbin.org/delay/3");
        logInfo("Delay request finished");
        auto r2 = getContent("http://google.com");
        logInfo("Google request finished");
    }

    setLogFormat(FileLogger.Format.threadTime, FileLogger.Format.threadTime);
    for(size_t i = 0; i < 3; i++)
        runTask(&taskMain);
}
```
output:
```
[F7EC2FAB:F7ECD7AB 2016.07.05 16:55:54.115 INF] Task created
[F7EC2FAB:F7ECD3AB 2016.07.05 16:55:54.116 INF] Task created
[F7EC2FAB:F7ED6FAB 2016.07.05 16:55:54.116 INF] Task created
[F7EC2FAB:F7ECD7AB 2016.07.05 16:55:57.451 INF] Delay request finished
[F7EC2FAB:F7ECD3AB 2016.07.05 16:55:57.464 INF] Delay request finished
[F7EC2FAB:F7ED6FAB 2016.07.05 16:55:57.474 INF] Delay request finished
[F7EC2FAB:F7ECD7AB 2016.07.05 16:55:57.827 INF] Google request finished
[F7EC2FAB:F7ECD3AB 2016.07.05 16:55:57.836 INF] Google request finished
[F7EC2FAB:F7ED6FAB 2016.07.05 16:55:57.856 INF] Google request finished
```

As you can see this generates 3 fibers, all in the same thread. The `getContent("http://httpbin.org/delay/3")` blocks the fiber. So the following `getContent("http://google.com")` is executed after the `http://httpbin.org/delay/3`.
But the Fibers execute in parallel. So all three requests to `http://httpbin.org/delay/3` happen at the same time.  The google requests then can also be executed in parallel.

Here's the output when using the same program with the std.socket backend:
```
[0CB25FEA:0CB207EA 2016.07.05 16:57:32.523 INF] Task created
[0CB25FEA:0CB207EA 2016.07.05 16:57:35.856 INF] Delay request finished
[0CB25FEA:0CB207EA 2016.07.05 16:57:36.195 INF] Google request finished
[0CB25FEA:0CB203EA 2016.07.05 16:57:36.195 INF] Task created
[0CB25FEA:0CB203EA 2016.07.05 16:57:39.525 INF] Delay request finished
[0CB25FEA:0CB203EA 2016.07.05 16:57:39.795 INF] Google request finished
[0CB25FEA:0CB29FEA 2016.07.05 16:57:39.795 INF] Task created
[0CB25FEA:0CB29FEA 2016.07.05 16:57:43.157 INF] Delay request finished
[0CB25FEA:0CB29FEA 2016.07.05 16:57:43.428 INF] Google request finished
```

ikod commented:
Great, this is exactly what we need. I'll put this as example in README.

I have several questions on building/testing:
1. Is it possible to configure dub.json so that it will fetch vibe.d as dependency only if user select vibed configuration?
2. This sequence fails for me:
```
Igors-MacBook-Pro:tmp igor$ mkdir xxx
Igors-MacBook-Pro:tmp igor$ cd xxx
Igors-MacBook-Pro:xxx igor$ git clone https://github.com/jpf91/dlang-requests.git
Cloning into 'dlang-requests'...
remote: Counting objects: 657, done.
remote: Compressing objects: 100% (38/38), done.
remote: Total 657 (delta 19), reused 0 (delta 0), pack-reused 613
Receiving objects: 100% (657/657), 391.60 KiB | 375.00 KiB/s, done.
Resolving deltas: 100% (358/358), done.
Checking connectivity... done.
Igors-MacBook-Pro:xxx igor$ cd dlang-requests/
Igors-MacBook-Pro:dlang-requests igor$ git checkout -b jpf91-vibe master
Switched to a new branch 'jpf91-vibe'
Igors-MacBook-Pro:dlang-requests igor$ git pull https://github.com/jpf91/dlang-requests.git vibe
From https://github.com/jpf91/dlang-requests
 * branch            vibe       -> FETCH_HEAD
Updating 550c130..9bba119
Fast-forward
 .travis.yml               |   4 ++
 README.md                 |  13 +++++
 dub.json                  |  27 +++++++---
 source/requests/base.d    |   4 +-
 source/requests/ftp.d     |  53 ++++++++++----------
 source/requests/http.d    |  92 +++++++++++-----------------------
 source/requests/streams.d | 393 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++----------------------------------------
 7 files changed, 377 insertions(+), 209 deletions(-)
Igors-MacBook-Pro:dlang-requests igor$ dub test
Unknown dependency: memutils
```
But just next run succeed without any problems:
```
Igors-MacBook-Pro:dlang-requests igor$ dub test
Unknown dependency: memutils
Igors-MacBook-Pro:dlang-requests igor$ dub test
Generating test runner configuration '__test__std__' for 'std' (library).
Performing "unittest" build using dmd for x86_64.
```
Can you reproduce this?

Thank you for your efforts!

ikod commented:
Looked at code, I see no other problems, will merge as soon as we solve problem with dependency.

jpf91 commented:
> Is it possible to configure dub.json so that it will fetch vibe.d as dependency only if user select vibed configuration?

AFAIK no: https://forum.rejectedsoftware.com/groups/rejectedsoftware.dub/thread/6962/ dub.json is already correct. It's a dub 'bug' that it still downloads vibe.d.

> This sequence fails for me:

sorry, I can't reproduce this (with dub 1.0.0 on linux).

ikod commented:
Merged, thanks a lot!

ikod commented on source/requests/streams.d:
```diff
@@ -634,122 +634,133 @@ public unittest {
-    SSL_METHOD* TLSv1_2_client_method();
-    SSL_METHOD* TLSv1_client_method();
-}
+version (sslLibs) {
```
Hello,

Do we really need "version(ssLibs)"?
Its presence require adding -version=sslLibs to each dmd call when we do not use dub:
```
igor$ cat y.d
#!/usr/bin/env rdmd -I./source librequests.a -L-lssl -L-lcrypto y.d
import std.stdio;
import std.algorithm;
import requests;
import std.experimental.logger;

void main()
{
    globalLogLevel(LogLevel.trace);
    Request rq;
    rq.verbosity = 3;
    rq.useStreaming = true;
    auto rs = rq.get("http://httpbin.org/stream/100");
    auto stream = rs.receiveAsRange;
    writeln(stream.joiner.count);
}
igor$ ./y.d
./source/requests/streams.d(1005): Error: undefined identifier 'SSLSocketStream'
Failed: ["dmd", "-I./source", "librequests.a", "-L-lssl", "-L-lcrypto", "-v", "-o-", "y.d", "-I."]
```
From the other side removing it makes no harm for vibed tests.
So can i remove it?

jpf91 commented:
> Do we really need "version(sslLibs)"?
> Its presence require adding -version=sslLibs to each dmd call when we do not use dub:

I recommend keeping it to avoid the openssl dependency when using vibe.d. IIRC vibe.d can also use other SSL libraries and then linking could fail if dlang-requests depends on OpenSSL.

But a simple solution for this problem is adding this code to stream.d:

```d
version (vibeD) {
}
else {
    version = sslLibs;
}
```

or simply replace `version(sslLibs) {` with `version(vibeD){} else {`

`"versions": ["sslLibs"]` could then get removed from dub.json as well.

ikod commented:
thanks!
